toolsets/cmake/tools.cmake

```diff
@@ -0,0 +1,311 @@
+    link_directories(${CMAKE_ARCHIVE_OUTPUT_DIRECTORY_RELWITHDEBINFO})
+  elseif (${CMAKE_BUILD_TYPE} STREQUAL "MinSizeRel")
+    link_directories(${CMAKE_ARCHIVE_OUTPUT_DIRECTORY_MINSIZEREL})
+  endif()
+endmacro()
+
+function(list_files ret dir)
+  file(GLOB _tmp
+    "${dir}/*.cpp"
+    "${dir}/*.hpp"
+    "${dir}/*.c"
+    "${dir}/*.h"
+    "${dir}/*.cc"
+    "${dir}/*.hh"
+    "${dir}/*.mm")
+  set(${ret} ${_tmp} PARENT_SCOPE)
+endfunction()
+
+macro(add_c_flag)
+  foreach(arg ${ARGN})
+    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} ${arg}")
+  endforeach()
+endmacro()
+
+macro(add_cxx_flag)
+  foreach(arg ${ARGN})
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${arg}")
+  endforeach()
+endmacro()
+
+macro(add_flags)
+  foreach(arg ${ARGN})
+    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} ${arg}")
+  endforeach()
+  foreach(arg ${ARGN})
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${arg}")
+  endforeach()
+endmacro()
+
+macro(add_project name)
+  message("-- Compiler: ${CMAKE_CXX_COMPILER_ID}")
+  message("-- Build for system ${CMAKE_SYSTEM_NAME}")
+  set_property(GLOBAL PROPERTY PROPERTY_PROJECT_NAME "${PROJECT_NAME}")
+  set(NISDK_OUTPUT_DIR "${CMAKE_CURRENT_SOURCE_DIR}/bin/$ENV{CMAKE_BUILD_ARCH}")
+  set(CMAKE_LIBRARY_OUTPUT_DIRECTORY ${NISDK_OUTPUT_DIR})
+  set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${NISDK_OUTPUT_DIR})
+  set(CMAKE_ARCHIVE_OUTPUT_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}/libs/$ENV{CMAKE_BUILD_ARCH}")
+
+  set(CMAKE_EXPORT_COMPILE_COMMANDS ON)  # enable compile_commands.json
+  set(CMAKE_CXX_STANDARD 14) # using c++ 14
+endmacro()
+
+
+macro(project_init)
+  message("-- Compiler: ${CMAKE_CXX_COMPILER_ID}")
+  message("-- Build for system ${CMAKE_SYSTEM_NAME}")
+  set_property(GLOBAL PROPERTY PROPERTY_PROJECT_NAME "${PROJECT_NAME}")
+  set(NISDK_OUTPUT_DIR "${CMAKE_CURRENT_SOURCE_DIR}/${CMAKE_PROJECT_NAME}/bin/$ENV{CMAKE_BUILD_ARCH}")
+  set(CMAKE_LIBRARY_OUTPUT_DIRECTORY ${NISDK_OUTPUT_DIR})
+  set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${NISDK_OUTPUT_DIR})
+  set(CMAKE_ARCHIVE_OUTPUT_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}/${CMAKE_PROJECT_NAME}/libs/$ENV{CMAKE_BUILD_ARCH}")
+
+
+  set(CMAKE_EXPORT_COMPILE_COMMANDS ON)  # enable compile_commands.json
+  set(CMAKE_CXX_STANDARD 14) # using c++ 14
+
+  if (CMAKE_SYSTEM_NAME STREQUAL "Darwin")
+    # We need to use libtool on apple to stop warnings
+    if(NOT CMAKE_LIBTOOL)
+      find_program(CMAKE_LIBTOOL NAMES libtool)
+    endif()
+
+    if(CMAKE_LIBTOOL)
+      set(CMAKE_LIBTOOL ${CMAKE_LIBTOOL} CACHE PATH "libtool executable")
+      message(STATUS "Found libtool - ${CMAKE_LIBTOOL}")
+      get_property(languages GLOBAL PROPERTY ENABLED_LANGUAGES)
+      foreach(lang ${languages})
+        set(CMAKE_${lang}_CREATE_STATIC_LIBRARY
+          "${CMAKE_LIBTOOL} -static -o <TARGET> -a -no_warning_for_no_symbols <LINK_FLAGS> <OBJECTS> ")
+      endforeach()
+    endif()
+  endif()
+
+  # Global flags
+  if (CMAKE_CXX_COMPILER_ID STREQUAL "Clang" OR CMAKE_CXX_COMPILER_ID STREQUAL "AppleClang")
+    add_c_flag(
```

Comment:
So this one is on me, but we should stop disabling warnings in the build scripts - including the .ham. We can actually put all of those in pragmas. Which is a lot less ugly... even though ideally our code should need those at all in an ideal world...